android/src/main/java/com/reactnativecommunity/viewpager/ReactViewPagerManager.java

```diff
@@ -59,186 +46,121 @@ protected ViewPager2 createViewInstance(@NonNull ThemedReactContext reactContext
     }
 
-    private void setCurrentItem(final ViewPager2 view, int selectedTab, boolean scrollSmooth) {
-        view.post(new Runnable() {
+    @Override
+    public void addView(ViewPager2 parent, View child, int index) {
+        final FragmentAdapter adapter = (FragmentAdapter) parent.getAdapter();
+        adapter.addReactView(child, index);
+        parent.post(new Runnable() {
             @Override
             public void run() {
-                view.measure(
-                        View.MeasureSpec.makeMeasureSpec(view.getWidth(), View.MeasureSpec.EXACTLY),
-                        View.MeasureSpec.makeMeasureSpec(view.getHeight(), View.MeasureSpec.EXACTLY));
-                view.layout(view.getLeft(), view.getTop(), view.getRight(), view.getBottom());
+                adapter.onAfterUpdateTransaction();
             }
         });
-        view.setCurrentItem(selectedTab);
-    }
-
-
-    @Override
-    public void addView(ViewPager2 parent, View child, int index) {
-        if (child == null) {
-            return;
-        }
-        ((FragmentAdapter) parent.getAdapter()).addFragment(child, index);
     }
 
     @Override
     public int getChildCount(ViewPager2 parent) {
-        return parent.getAdapter().getItemCount();
+        return ((FragmentAdapter) parent.getAdapter()).getReactChildCount();
     }
 
-
     @Override
     public View getChildAt(ViewPager2 parent, int index) {
-        return ((FragmentAdapter) parent.getAdapter()).getChildViewAt(index);
+        return ((FragmentAdapter) parent.getAdapter()).getReactChildAt(index);
     }
 
     @Override
-    public void removeView(ViewPager2 parent, View view) {
-        ((FragmentAdapter) parent.getAdapter()).removeFragment(view);
-    }
-
-
-    public void removeAllViews(ViewPager2 parent) {
-        FragmentAdapter adapter = ((FragmentAdapter) parent.getAdapter());
-        adapter.removeAll();
+    public void removeViewAt(ViewPager2 parent, int index) {
+        final FragmentAdapter adapter = (FragmentAdapter) parent.getAdapter();
+        adapter.removeReactViewAt(index);
+        parent.post(new Runnable() {
+            @Override
+            public void run() {
+                adapter.onAfterUpdateTransaction();
+            }
+        });
     }
 
+    @Nullable
     @Override
-    public void removeViewAt(ViewPager2 parent, int index) {
-        FragmentAdapter adapter = ((FragmentAdapter) parent.getAdapter());
-        adapter.removeFragmentAt(index);
+    public Map<String, Object> getExportedCustomDirectEventTypeConstants() {
+        return MapBuilder.<String, Object>of(
+                PageScrollEvent.EVENT_NAME, MapBuilder.of("registrationName", "onPageScroll"),
+                PageSelectedEvent.EVENT_NAME, MapBuilder.of("registrationName", "onPageSelected"));
     }
 
     @Override
-    public boolean needsCustomLayoutForChildren() {
-        return true;
+    protected void onAfterUpdateTransaction(@NonNull ViewPager2 view) {
```

Comment:
RN calls this method after calling all applicable `@ReactProp` setters.  RN would not otherwise call `onAfterUpdateTransaction()` in the adapter, as it does not have knowledge of the adapter.